### PR TITLE
BANNO.LOANPAYMENT.SKIP.V1 1.1.2 Update

### PR DESCRIPTION
**Pull request by tkainz**
* Modified CFG Letter file read error handling
* Corrected loan disqualification check loop
* Reworked param LF read procedure
* Other code cleanup

### CODE REVIEW

**Review by jucarson — COMMENTED** [review]
Code looks good. 5 comments in my review are just typos. And I have this on the worksheet:

- Any reason why we went PDF instead of DOCX? Seems harder to fill out. Maybe just me.
- Approval Code Exclusions has a red "Blank=". Not sure if it goes there.
- Max number of late payments description should say "..of late payments that has been met or exceeded"
- Eligibile share types for "feeS" typo

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 1]
```diff
@@ -150,17 +155,21 @@ DEFINE
...
+ INELIGIBLECODELOANWARNING = 2  [has loan warning]
+ INELIGIBLECODEMAXSKIPS    = 3  [has used max skips already]
+ INELIGIBLECODETIMESKIP    = 4  [not enough time between skips]
+ INELIGIBLECODEACCTWARNING = 5  [has account warning]
+ INELIGIBLECODESERVICECODE = 6  [does not have required svc code]
+ INELIGIBLECODEMINPAYMENT  = 7  [min payment requirement not met]
+ INELIGIBLECODEMAXPAYMENT  = 8  [max payment amount exceeded]
+ INELIGIBLECODETIMEOPEN    = 9  [loan not opened for sufficient time]
+ INELIGIBLECODEPASTDUE     = 10 [past due grace days exceeded]
+ INELIGIBLECODEFREQ        = 11 [has ineligibility payment freq]
```
Comment (jordan-keenan) [inline loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:169]: I think "has ineligibility", should be "has ineligible"

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 2]
```diff
@@ -8,6 +8,7 @@
 * than 126 characters. Each parameter setting consists of a parameter
 * name followed by a colon and the parameter value
 *  (for example: 'CT: 1')
+*
 * Where instructions call for entering a "range of values", enter single
 * or group values seperated by commas - ex. 1,10-20,30,33.
```
Comment (jordan-keenan) [inline loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:13]: should be "separated"

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 3]
```diff
@@ -24,236 +25,232 @@
...
-*   example: DQA:2 means that the loan will be disqualified for a skip once the
-*            late payment count is 2 or greater for category DQA (11-30 DQ days)
-*  DATA CONSTRAINTS: Numeric value for each category to indicate the number
-*                    of late payments for that category when the loan becomes
-*                    ineligible. '0' indicates that category will not be taken into
-*                    consideration when assessing eligibility
-*  DEFAULT: '0' for each category (late payments not taken into consideration)
+* MAXIMUM NUMBER OF LATE PAYMENTS
+* The number of late payments, by category, which would disqualify a
+* loan. If the number of late payments in any one catogory have been
```
Comment (jordan-keenan) [inline loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:182]: should be "category"

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 4]
```diff
@@ -267,48 +264,113 @@ DDA11:M
...
+*  DEFAULT:               Skip Payment Fee
+FC:Skip Payment Fee
+*
+* FEE COMMENT OPTION
+* Represents the description to post a loan comment transaction for
+* Reg Z fee types paid from a share for a loan transaction. If set to 0
+* or left blank, the comment will be added to the share transaction and
+* if set to 1, the comment will be added to the loan transaction.
+*  LINE MUST BEGIN WITH: "OA"
+*  DATA FORMAT:           numberic
```
Comment (jordan-keenan) [inline loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:332]: change to "numeric"

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 5]
```diff
@@ -267,48 +264,113 @@ DDA11:M
...
+*  DATA CONSTRAINTS:      >=0 & <=1
+*  DEFAULT:               0
+OA:0
+*
+* SUB SOURCE CODE FOR FEE
+* The subsourcecode to be assigned to the fee transaction and to appear
+* in the transaction's sub source code field.
+*  LINE MUST BEGIN WITH: "SS"
+*  DATA FORMAT:           numberic
```
Comment (jordan-keenan) [inline loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:341]: change to "numeric"

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 6]
```diff
@@ -24,236 +25,232 @@
...
-*    skip first, 7:biweekly skip last, 8:biweekly, 9:weekly, 11:weekly skip first
-*   12:weekly skip last, 13:bimonthly
-*  Settings for frequencies not offered (see "Eligible Payment Frequencies" above)
-*  will be ignored or they can be left out completely.
-*  DATA CONSTRAINTS: "M" or "F"
-*  DEFAULT: "M" for each frequency (1 month/28 days) due date advancement.
+*************************** PROGRAM CRITERIA ***************************
+* These are the options impacting program functionality and skip payment
+* processing. Once set and the program is operational, these settings
+* should not be changed without first carefuly evaluating the potential
```
[the remark this answers is not in the document]
Comment (tkainz) [reply loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:206]: Corrected

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG [hunk 7]
```diff
@@ -267,48 +264,113 @@ DDA11:M
...
+* or left blank, the comment will be added to the share transaction and
+* if set to 1, the comment will be added to the loan transaction.
+*  LINE MUST BEGIN WITH: "OA"
+*  DATA FORMAT:           numberic
+*  DATA CONSTRAINTS:      >=0 & <=1
+*  DEFAULT:               0
+OA:0
+*
+* SUB SOURCE CODE FOR FEE
+* The subsourcecode to be assigned to the fee transaction and to appear
```
[the remark this answers is not in the document]
Comment (tkainz) [reply loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:338]: Corrected

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 8]
```diff
@@ -150,17 +155,21 @@ DEFINE
...
+ INELIGIBLECODEMINPMTCOUNT = 12 [minimum payment countnot met]
```
[the remark this answers is not in the document]
Comment (tkainz) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:170]: Corrected

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 9]
```diff
@@ -785,49 +786,75 @@ PROCEDURE READCONFIGFILESETTINGS
...
-[ Terms & conditions]
-       IF SEGMENT(LFLINE,1,3)="TC:" AND TERMSLINESCOUNT<TERMSMAXLINES THEN
+[The text in front of the colon is the parameter and after the colon is
+ the parameter setting.]
+       POS=CHARACTERSEARCH(LFLINE,":")
+       TEXT1=SEGMENT(LFLINE,1,POS-1)
+       CALL TEXTNLTS
+       TMPPARAM=TEXT1
+       TEXT1=SEGMENT(LFLINE,POS+1,LENGTH(LFLINE))
+[Trim leading and trailing spaces from the paramater value unless it's
```
[the remark this answers is not in the document]
Comment (tkainz) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:804]: Corrected

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW (hunk 1 above)
Comment (tkainz) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:169]: Changed

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG (hunk 2 above)
Comment (tkainz) [reply loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:13]: Corrected

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG (hunk 3 above)
Comment (tkainz) [reply loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:182]: Corrected

File: loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG (hunk 5 above)
Comment (tkainz) [reply loan-payment-skip/LETTERSPECS/BANNO.LOANPAYMENT.SKIP.CFG:341]: Corrected

**Comment by tkainz** [comment]
Spelling type-os have been addressed
Worksheet changed from PDF to docx

**Review by jordan-keenan — CHANGES_REQUESTED** [review]
Hi Tom,
 I was meaning to send this before your update came through this morning.
Can you take a look at the following and likely make changes to the worksheet?
I'm guessing you took care of the spelling things we noticed for clean up.

First thank you for making the worksheet a word document, as all CU's could use the keyboard
to fill in the answers with MS Word, where with PDF some may not want to bother figuring out how to enter text mode to add  values.

Can you make changes to the worksheet?
I don't think including the Letters used by the Letter File with a blank after
them for each field such as "LT:" is a good idea. It would seem to encourage
CU's to leave fields blank more often. Thad being said if you are going to do
that, please define what blank means in the description for Every Parameter
Description, and say Blank is not a valid value in each one where blank is not valid. 

I don't mean to be disagreeable, but I think now is a good time to talk
about problems this approach may introduce.

 I'm also going to look over the logic, which is probably fine, before my next review of this.

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 10]
```diff
@@ -150,17 +155,21 @@ DEFINE
...
+ INELIGIBLECODEACCTWARNING = 5  [has account warning]
+ INELIGIBLECODESERVICECODE = 6  [does not have required svc code]
+ INELIGIBLECODEMINPAYMENT  = 7  [min payment requirement not met]
+ INELIGIBLECODEMAXPAYMENT  = 8  [max payment amount exceeded]
+ INELIGIBLECODETIMEOPEN    = 9  [loan not opened for sufficient time]
+ INELIGIBLECODEPASTDUE     = 10 [past due grace days exceeded]
+ INELIGIBLECODEFREQ        = 11 [has ineligibility payment freq]
+ INELIGIBLECODEMINPMTCOUNT = 12 [minimum payment countnot met]
+ INELIGIBLECODEAPPROVALCODE= 13 [has invalid approval code]
+ INELIGIBLECODEDQPAYMENTS  = 14 [has excessive DQ paymrnt history]
```
[the remark this answers is not in the document]
Comment (tkainz) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:172]: Corrected

**Review by jucarson — APPROVED** [review]
Looks Good Tom!

**Review by jordan-keenan — CHANGES_REQUESTED** [review]
Thank you for all the cleanup Tom.
I see the changes you made we recommended last time look good.

Since the goal now, per the email chain and changed up worksheet that now includes Parameter Tags, is if the CU puts blank next to the Parameter Tag then
give them the default value, I have found the following issues:

In the New Config worksheet, Loan Payment Skip Config Worksheet.docx:
For Parameter "Maximum Number of Late Payments", in CU Response "DQA" needs a ":"
after it.

For Parameter "Fee Comment" in Parameter Description in the
"Valid Values:" section I recommend you say it can be a max of 40 characters.
Also, since on the worksheet now leaving blank after the tag
"FC:" means blank, I suggest you add the following concept
somewhere to the Parameter Description, "Example: Skip Payment Fee".

…

For the Letter File BANNO.LOANPAYMENT.SKIP.CFG
I noticed the comments for Parameter "Fee Comment" says DEFAULT : Skip Payment Fee,
which is not true; it is blank in the code.
 I suggest changing the description for default to blank in the CFG.
I'm not sure if there is a good way to make a blank value default to 
"Skip Payment Fee" code wise, as some rare customers may actually want it blank.

…

For Code, BANNO.LOANPAYMENT.SKIP.V1.POW: 
I understand that we did not write the original code in this department.
However; due to the config worksheet change to blank after a tag means use default (unless stated otherwise) the following code changes would need to be made to make the 
PowerOn work that way.

You can either move condition "AND TMPPARAMVAL<>" to the top of the Letter File Line reading Logic just before starting to check the value of each "TMPPARAM", and set any default values properly in the Initialize data section that would be affected by that (ELEGIBLEFREQUENCIES for example) OR
You can change each of the following;  flowing the suggestions below:

[ Share types ] I suggest for code consistency like loan types add  IF TMPPARAMVAL="" THEN TMPPARAMVAL="ALL" before LELISTINPUT
.
       ELSE IF TMPPARAM="ST" THEN
        DO
         LELISTINPUT=TMPPARAMVAL

[ Fee amount(s) ] is missing AND TMPPARAMVAL<>" after the Else If clause here
.
       ELSE IF TMPPARAM="FA" THEN
        DO
         FEEAMOUNTS(VALUE(SEGMENT(TMPPARAMVAL,1,2)))=
                    MONEY(VALUE(SEGMENT(TMPPARAMVAL,4,LENGTH(TMPPARAMVAL))))
        END

[ Max skips per year ] is missing AND TMPPARAMVAL<>" after the Else If clause here
.
       ELSE IF TMPPARAM="PS" THEN
        DO
         MAXSKIPS=VALUE(TMPPARAMVAL)
        END

[ Min months since last skip ] is missing AND TMPPARAMVAL<>" after the Else If clause her
.
       ELSE IF TMPPARAM="MM" THEN
        DO
         MINMONTHSINCE=VALUE(TMPPARAMVAL)
        END

.
While this is not a bug, LOANSEASON should probably be set to 0 in the
Initialize section.

Thank you

**Comment by tkainz** [comment]
@jordan-keenan Your requested changes have been completed.

**Review by jordan-keenan — APPROVED** [review]
Code Changes look Good Tom. Thanks for doing that!
I only have one suggestion, now that the CFG the way the code works,  for "FC" fee comments, I suggest, like how the worksheet gives a hint, that you also add
something line "Example “Skip Payment Fee”  " somewhere in the FEE COMMENT Parameter comments section in BANNO.LOANPAYMENT.SKIP.CFG

**Review by jordan-keenan — APPROVED** [review]
Code Looks good!
Thank you

**Pull request merged**